apps/openmw/mwsound/volumesettings.cpp

```diff
@@ -0,0 +1,56 @@
+#include "volumesettings.hpp"
+
+#include <components/settings/settings.hpp>
+
+#include <algorithm>
+
+namespace MWSound
+{
+    namespace
+    {
+        float clamp(float value)
+        {
+            return std::max(0.0f, std::min(1.0f, value));
+        }
+    }
+
+    VolumeSettings::VolumeSettings()
+        : mMasterVolume(clamp(Settings::Manager::getFloat("master volume", "Sound"))),
+          mSFXVolume(clamp(Settings::Manager::getFloat("sfx volume", "Sound"))),
+          mMusicVolume(clamp(Settings::Manager::getFloat("music volume", "Sound"))),
+          mVoiceVolume(clamp(Settings::Manager::getFloat("voice volume", "Sound"))),
+          mFootstepsVolume(clamp(Settings::Manager::getFloat("footsteps volume", "Sound")))
+    {
+    }
+
+    float VolumeSettings::getVolumeFromType(Type type) const
+    {
+        float volume = mMasterVolume;
+
+        switch(type)
+        {
+            case Type::Sfx:
+                volume *= mSFXVolume;
+                break;
+            case Type::Voice:
+                volume *= mVoiceVolume;
+                break;
+            case Type::Foot:
+                volume *= mFootstepsVolume;
+                break;
+            case Type::Music:
+                volume *= mMusicVolume;
+                break;
+            case Type::Movie:
+            case Type::Mask:
+                break;
+        }
+
+        return volume;
+    }
+
+    void VolumeSettings::update()
+    {
+        *this = VolumeSettings();
```

Comment:
Are you sure that such approach is going to work everywhere?

Reply:
This is a call of an assignment operator. I can rewrite it as:
```c++
void VolumeSettings::update()
{
    operator=(VolumeSettings());
}
```
I don't see why this may cause problems.

I didn't want to use fill method, because it delays initialization and it's easier to miss a field.